torch/nn/utils/parametrizations.py

```diff
@@ -1,10 +1,273 @@
+    base: Tensor
+
+    def __init__(self, weight, parametrization: str = None, *, extra_memory=True) -> None:
+        super().__init__()
+
+        # Note [Householder complex]
+        # For complex tensors, it is not possible to compute the tensor `tau` necessary for
+        # linalg.householder_product from the reflectors.
+        # To see this, note that the reflectors have a shape like:
+        # 0 0 0
+        # * 0 0
+        # * * 0
+        # which, for complex matrices, give n(n-1) (real) parameters. Now, you need n^2 parameters
+        # to parametrize the unitary matrices. Saving tau on its own does not work either, because
+        # not every combination of `(A, tau)` gives a unitary matrix, meaning that if we optimise
+        # them as independent tensors we would not maintain the constraint
+        # An equivalent reasoning holds for rectangular # matrices
+        if weight.is_complex() and parametrization == "householder":
+            raise ValueError("The householder parametrization does not support complex tensors.")
+
+        # We could implement this for 1-dim tensors as the maps on the sphere
+        # but I believe it'd bite more people than it'd help
+        if weight.ndim < 2:
+            raise ValueError("Expected a matrix or batch of matrices. "
+                             f"Got a tensor of {weight.ndim} dimensions.")
+
+        self.shape = weight.shape
+        self.parametrization = parametrization
+        if extra_memory:
+            self.register_buffer("base", None)
+
+    def forward(self, X: torch.Tensor) -> torch.Tensor:
+        if X.shape != self.shape:
+            raise ValueError(f"Expected a matrix or batch of matrices of shape {self.shape}. "
+                             f"Got a tensor of shape {X.shape}.")
+
+        n, k = X.size(-2), X.size(-1)
+        transposed = n < k
+        if transposed:
+            X = X.transpose(-2, -1)
+            n, k = k, n
+        # Here n > k and X is a tall matrix
+        if self.parametrization == "matrix_exp" or self.parametrization == "cayley":
+            # We just need n x k - k(k-1)/2 parameters
+            X = X.tril()
+            if n != k:
+                # Embed into a square matrix
+                X = torch.cat([X, X.new_zeros(n, n - k).expand(*X.shape[:-2], -1, -1)], dim=-1)
+            A = X - X.transpose(-2, -1).conj()
+            # A is skew-symmetric (or skew-hermitian)
+            if self.parametrization == "matrix_exp":
+                Q = torch.matrix_exp(A)
+            elif self.parametrization == "cayley":
+                # Computes the Cayley retraction (I+A/2)(I-A/2)^{-1}
+                Id = torch.eye(n, dtype=A.dtype, device=A.device)
+                Q = torch.linalg.solve(torch.add(Id, A, alpha=-0.5), torch.add(Id, A, alpha=0.5))
+            # Q is now orthogonal (or unitary) of size (..., n, n)
+            if n != k:
+                Q = Q[..., :k]
+            # Q is now the size of the original tensor
```

Comment:
nit: of X at this point. As in the transposed case it wouldn't match the original Tensor?